src/core/websocket/ws.ts

```diff
@@ -452,35 +500,106 @@ export class ReconnectingWebSocket {
+    }
 
-        if (event.code === 1000) {
-            this._shouldReconnect = false;
-        }
+    /**
+     * Core handler for all errors. It first standardizes the error input using `_adaptError`,
+     * then manages disconnection, dispatches the standardized error event, and initiates reconnection.
+     * @param rawErrorOrEventInput - The error data from any internal source.
+     */
+    private _handleError = (rawErrorOrEventInput: any): void => {
+        this._debug("Raw error/event received by _handleError, passing to _adaptError:", rawErrorOrEventInput);
+
+        const adaptedError = this._adaptError(rawErrorOrEventInput);
+        const { error, message } = adaptedError;
 
-        if (this._shouldReconnect) {
+        // Close with abnormal closure code so _handleClose will reconnect if there's a socket
+        this._disconnect(ABNORMAL_CLOSURE_CODE, message);
+
+        this._debug("Processed error. Dispatching Event:", message, error);
+        if (this.onerror) this.onerror(adaptedError);
+
+        this._debug("exec error listeners");
+        this._listeners.error.forEach((listener) => this._callEventListener(adaptedError, listener));
+
+        // If we never even got a ws instance, _disconnect returned early so we need to explicitly retry here
+        if (!this._ws && this._shouldReconnect) {
+            this._debug("Error before socket created; scheduling initial reconnect.");
             this._connect();
         }
+    };
 
-        if (this.onclose) {
-            this.onclose(event);
+    /**
+     * Handles WebSocket close events, either native or synthesized internally.
+     *
+     * Adapts the event if necessary, manages state, dispatches standardized close events,
+     * and initiates reconnection if appropriate.
+     *
+     * @param eventInfo - Either an internal Events.CloseEvent or a native DOM CloseEvent.
+     */
+    private _handleClose = (event: Events.CloseEvent | CloseEvent) => {
+        this._debug("Raw close event received by _handleClose:", event);
+
+        // Adapt event if event is an instance of the (browser) native CloseEvent
+        const isNativeCloseEvent = typeof CloseEvent !== "undefined" && event instanceof CloseEvent;
+        const adaptedEvent: Events.CloseEvent = isNativeCloseEvent
+            ? new Events.CloseEvent(event.code, event.reason, this)
+            : (event as Events.CloseEvent);
+
+        // Clean up state
+        this._clearTimeouts();
+        this._ws = undefined;
+        this._connectLock = false;
+
+        // Determine reconnection intent
+        let attemptReconnect = this._shouldReconnect;
+        if (this._closeCalled) {
+            attemptReconnect = false;
+            this._debug("Reconnection stopped: RWS.close() was called.");
+        } else if (adaptedEvent.code === 1000) {
+            attemptReconnect = false;
+            this._debug("Reconnection stopped: Received close code 1000 (intentional server close).");
```

Comment:
For me it's clearer without the local variable
```suggestion
        if (this._closeCalled) {
            this._shouldReconnect = false;
            this._debug("Reconnection stopped: RWS.close() was called.");
        } else if (adaptedEvent.code === 1000) {
            this.shouldReconnect = false;
            this._debug("Reconnection stopped: Received close code 1000 (intentional server close).");
```